FEAT directory support

infocortex opened the pull request:
works well with original example.  importing the converter class seems preferable to calling the executable, and gives us control over the exception handling in our process.

todo:
 - testing
 - etc

chrisfilo commented on neurovault/apps/statmaps/utils.py:
```diff
@@ -281,16 +283,103 @@ def populate_nidm_results(request,collection):
+
+
+
+
```
?

infocortex replied:
just some accident whitespace in the commit.

chrisfilo commented:
Excellent. Please use the data I've sent you for test, add dependencies to requirements.txt and update puppet.

chrisfilo commented on neurovault/apps/statmaps/views.py:
```diff
@@ -350,6 +352,11 @@ def upload_folder(request, collection_cid):
                             request.FILES['file'] = f
                             populate_nidm_results(request,collection)
                             continue
+                        elif fnmatch(f.name,'*.feat.zip'):
```
This is not how I imaging people will use it. FEAT directories will be given to us either through "Upload directory" mechanism or "Upload archive" (where we should not expect special .feat.zip extension). Thus we need to try to detect if the uploaded folder structure is FEAT or not.

infocortex replied:
OK, that should be doable- the directory structure of the uploaded folder is lost, but it can be reconstructed with the `paths[]` post variable.

One thing that might need refinement, is a specification for conclusively detecting the directory types.  I can start with some reasonable assumptions for that, and I'm sure we can clarify it later if necessary.

infocortex commented:
Hi, I made a test script [check_feat_dirs.py](https://github.com/infocortex/NeuroVault/blob/b303b70c853317ed1dbaa8ddba671b2f65281c3c/scripts/check_feat_dirs.py) to test the nidm-results_fsl exporter against the whole large test dataset chris shared.

At the moment, nidm-results_fsl isn't successfully parsing the data, I'm seeing this error for each subject:
```python
found feat directory at '/vagrant/FEAT/ds116/group/model001/task001/cope001.gfeat/cope1.feat'
Unable to parse the FEAT directory:
Traceback (most recent call last):
  File "/opt/nv_env/NeuroVault/scripts/check_feat_dirs.py", line 23, in <module>
    fslnidm.parse()
  File "/opt/nv_env/lib/nidm-results_fsl/fsl_exporter/fsl_exporter.py", line 66, in parse
    self._get_coordinate_system()
  File "/opt/nv_env/lib/nidm-results_fsl/fsl_exporter/fsl_exporter.py", line 509, in _get_coordinate_system
    custom_space = (self._search_in_fsf(custom_re) == "1")
  File "/opt/nv_env/lib/nidm-results_fsl/fsl_exporter/fsl_exporter.py", line 540, in _search_in_fsf
    info = info_found.group('info')
AttributeError: 'NoneType' object has no attribute 'group'
```

@cmaumet , perhaps you might have an idea on what is amiss? 

(this is on the `convert_to_turtle` branch that works with the original example.)

chrisfilo commented:
It seems that there is something "missing" in the .fsf file. I'm afraid
@cmaumet is on holiday. Could you try to figure out which line the
converter is missing in the new fsf file? A good starting point would be to
add a more meaningful exception (that include that filename and the regexp)
here:
https://github.com/cmaumet/nidm-results_fsl/blob/convert_to_turtle/fsl_exporter/fsl_exporter.py#L538

On Tue Jan 13 2015 at 2:37:39 AM Gabriel Rivera <notifications@github.com>
wrote:

> Hi, I made a test script check_feat_dirs.py
> <https://github.com/infocortex/NeuroVault/blob/b303b70c853317ed1dbaa8ddba671b2f65281c3c/scripts/check_feat_dirs.py>
> to test the nidm-results_fsl exporter against the whole large test dataset
> chris shared.
>
> At the moment, nidm-results_fsl isn't successfully parsing the data, I'm
> seeing this error for each subject:
>
> found feat directory at '/vagrant/FEAT/ds116/group/model001/task001/cope001.gfeat/cope1.feat'
> Unable to parse the FEAT directory:
> Traceback (most recent call last):
>   File "/opt/nv_env/NeuroVault/scripts/check_feat_dirs.py", line 23, in <module>
>     fslnidm.parse()
>   File "/opt/nv_env/lib/nidm-results_fsl/fsl_exporter/fsl_exporter.py", line 66, in parse
>     self._get_coordinate_system()
>   File "/opt/nv_env/lib/nidm-results_fsl/fsl_exporter/fsl_exporter.py", line 509, in _get_coordinate_system
>     custom_space = (self._search_in_fsf(custom_re) == "1")
>   File "/opt/nv_env/lib/nidm-results_fsl/fsl_exporter/fsl_exporter.py", line 540, in _search_in_fsf
>     info = info_found.group('info')AttributeError: 'NoneType' object has no attribute 'group'
>
> @cmaumet <https://github.com/cmaumet> , perhaps you might have an idea on
> what is amiss?
>
> (this is on the convert_to_turtle branch that works with the original
> example.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NeuroVault/NeuroVault/pull/102#issuecomment-69725446>.
>

infocortex commented:
I was able to get it working with a few small changes: https://github.com/infocortex/nidm-results_fsl/commit/e332ddc9002dcb5f2927df73d59031c94080c9f5?w=1

- alternateReference_yn setting was missing in design.fsf
- deleting calculated_sigmasquareds.nii.gz was failing because it seems to be needed again later
- directories such as cope001 with leading zeros was a problem in the contrast number parsing

So, we should be set- all though I guess we might see more variations in the format parsing with other datasets.

chrisfilo commented:
Excellent - it would be great if you could send them a PR.

On Tue Jan 13 2015 at 11:25:25 PM Gabriel Rivera <notifications@github.com>
wrote:

> I was able to get it working with a few small changes:
> infocortex/nidm-results_fsl@e332ddc?w=1
> <https://github.com/infocortex/nidm-results_fsl/commit/e332ddc9002dcb5f2927df73d59031c94080c9f5?w=1>
>
>    - alternateReference_yn setting was missing in design.fsf
>    - deleting calculated_sigmasquareds.nii.gz was failing because it
>    seems to be needed again later
>    - directories such as cope001 with leading zeros was a problem in the
>    contrast number parsing
>
> So, we should be set- all though I guess we might see more variations in
> the format parsing with other datasets.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NeuroVault/NeuroVault/pull/102#issuecomment-69878477>.
>

infocortex commented:
OK, I think we're all ready to deploy:

- Unit tests: fed4fa6
- Added a modal processing dialog during upload, since the FEAT directories take quite some time: 0a05e9c
- Puppet installation of FEAT libaries: https://github.com/NeuroVault/neurovault_puppet/pull/15
- Modified changes to `nidm-results_fsl` exporter (I'll PR shortly): https://github.com/infocortex/nidm-results_fsl/compare/nv_testdata_fixes

infocortex commented:
PR sent for nidm-results_fsl: https://github.com/cmaumet/nidm-results_fsl/pull/1.

chrisfilo commented:
Despite the fact that test pass when I try to upload any of the example data .zip files (from openfmri or the one from Camille) I get redirected to /collections/13/upload_folder. What is more when I go back to the collection there is no sign of any new NIDM Results records. Can you replicate this?

chrisfilo commented:
It also needs to work when "Upload folder" (uncompressed) is being used.

infocortex commented:
Sorry, that was recently introduced, fixed in  0a05e9c.

chrisfilo commented:
After those fixes I get redirected back to the collection page (which is good), but no NIDM files are added (when I try with the two openfmri examples). When I use Camilles example it is treated like any other .zip folder (in other words all nifti files are added to the collection). Could you improve the tests or try those three files in order to find and fix the bug?

infocortex commented:
These are all working for me, as zips and folders.

By the openfmri examples, you mean are the ones already added as the test files?

Not sure what's going wrong.  Can you send me the failing zips?

infocortex commented:
Could your javascript still be cached, perhaps?

chrisfilo commented:
I tried with fresh vagrant instance and cleared cache (using a plugin as well as icognito mode). This is the behaviour I observe:

- for ds017.zip and ds005.feat.zip from the test_date folder I get redirected to the collection page after the upload with no new items in the collection
- if I upload corresponding folders instead of .zip files for those two datasets the behaviour is the same but I get a "Processing upload" dialog box. Still no new items in the collection when it's done processing
- if I use Camilles example ex_fmri_one_contrast.feat.zip nifti files are added to a collection as if it was a normal zip file. No new NIDM records.

infocortex commented:
OK, these issues should be resolved.
- uwsgi issue fixed in  2405685 and https://github.com/NeuroVault/neurovault_puppet/commit/e3b9190dfbd7edec9d4a261752939e4aaf1d494a
- I also added a error notification for failures in folder and zip uploads: 5c40704

infocortex commented:
True in, but it shouldn't be a problem with the path being explicitly added in our app.

Also, if the provconvert binary fails for some other reason, the subsequent exception in the nidm parsing will now throw a visible exception.  (nidm.ttl won't exist in a usable state).

chrisfilo commented:
I just tested it - without fixing the path it still fails silently (instead of rising an exception). After fixing the path it works (for openfmri examples). It still fails to recognise Camille's example as FEAT dir.

chrisfilo commented:
I've updated nidmresults to raise an exception if the conversion fails https://github.com/incf-nidash/nidmresults/pull/2 . To test it I on purpose removed provconvert from path and now I am getting this error:

	ValueError at /collections/5/upload_folder
	The view neurovault.apps.statmaps.views.upload_folder didn't return an HttpResponse object. It returned None instead.

infocortex commented:
Sounds good, catching at the subprocess call is definitely preferable.

The related error handling is fixed in  8225839.  I tested normal error cases, and also a missing provconvert binary, for zips and folders.

> It still fails to recognise Camille's example as FEAT dir.
Make sure you aren't still using the original copy of that file that had no contents in the logs/ or stats/ directory, that is likely the issue.
